src/libs/auth.js

```diff
@@ -327,32 +327,50 @@ authStore.subscribe(withErrorReporting('Error checking registration', async (sta
-    if (_.isNull(tosDetails)) {
-      const termsOfService = {
-        isGracePeriodEnabled: undefined,
-        currentVersion: undefined,
-        userAcceptedVersion: undefined,
-        userAcceptedTos: false,
-        userNeedsToAcceptTos: true,
-      }
-      authStore.update(state => ({ ...state, termsOfService }))
-    } else {
-      const userAcceptedTos = !_.isUndefined(tosDetails.currentVersion) ?
-        tosDetails.currentVersion === tosDetails.userAcceptedVersion : undefined
-      const userNeedsToAcceptTos = !_.isUndefined(userAcceptedTos) ?
-        (!userAcceptedTos && !tosDetails.isGracePeriodEnabled) || _.isUndefined(tosDetails.userAcceptedVersion) : false
-      const termsOfService = {
-        isGracePeriodEnabled: tosDetails.isGracePeriodEnabled,
-        currentVersion: tosDetails.currentVersion,
-        userAcceptedVersion: tosDetails.userAcceptedVersion,
-        userAcceptedTos,
-        userNeedsToAcceptTos,
-      }
-      authStore.update(state => ({ ...state, termsOfService }))
-    }
+    const termsOfService = parseToSDetails(tosDetails)
+    authStore.update(state => ({ ...state, termsOfService }))
   }
 }))
 
+const parseToSDetails = tosDetails => {
+  if (_.isNull(tosDetails)) {
+    return {
+      isGracePeriodEnabled: undefined,
+      currentVersion: undefined,
+      userAcceptedVersion: undefined,
+      userAcceptedTos: false,
+      userNeedsToAcceptTos: true,
+    }
+  } else {
+    // IF user has accepted latest version of ToS
+    // THEN let them use Terra, do not show them the ToS popup
+    // ELSE IF user accepted ANY prior version of ToS and Grace Period is enabled
+    // THEN let them use Terra, do show them the ToS popup
+    // ELSE                ///////// IF user has no accepted ToS version OR Grace Period is disabled
+    // THEN they can't use Terra until you accept ToS
+
+    // Has the user ever accepted any version of the ToS?
+    const userHasAcceptedAnyToSVersion = !_.isUndefined(tosDetails.userAcceptedVersion)
+    // Has the user accepted the latest/current version of the ToS?
+    const userAcceptedLatestTos = !_.isUndefined(tosDetails.currentVersion) ?
+      tosDetails.currentVersion === tosDetails.userAcceptedVersion : false
+    // Will we permit the user to use Terra under the ToS grace period?
+    const userOperatingUnderGracePeriod = tosDetails.isGracePeriodEnabled && userHasAcceptedAnyToSVersion
+    // Should we display the "Updated ToS" alert/pop-up to the user to ask them to accept the _latest_ version of ToS
+    const showToSPopup = !userAcceptedLatestTos
+
+    const userCanUseTerra = userAcceptedLatestTos || userOperatingUnderGracePeriod
+
+    return {
+      isGracePeriodEnabled: tosDetails.isGracePeriodEnabled,
+      currentVersion: tosDetails.currentVersion,
+      userAcceptedVersion: tosDetails.userAcceptedVersion,
+      userAcceptedTos: userCanUseTerra,
+      userNeedsToAcceptTos: showToSPopup,
```

Comment:
Could variable names match keys of the returned object to avoid having two different names for the same values within this function?

```suggestion
    const userNeedsToAcceptTos = !userAcceptedLatestTos

    const userAcceptedTos = userAcceptedLatestTos || userOperatingUnderGracePeriod

    return {
      isGracePeriodEnabled: tosDetails.isGracePeriodEnabled,
      currentVersion: tosDetails.currentVersion,
      userAcceptedVersion: tosDetails.userAcceptedVersion,
      userAcceptedTos, // if true, user can use Terra
      userNeedsToAcceptTos, // if true, show ToS alert to ask them to accept the _latest_ version of ToS
```

Reply:
agreed, i think we should do that in a followup pr though